### ISSUE TITLE
BN slows down double-backprop enormously

### ISSUE THREAD

**Issue by cjsg** [pytorch/pytorch#3076]
When using a ConvNet without batchnorm layers, the optimization with a gradient penalty takes approximately 5 times longer than without the gradient penalty. This ratio 1:5 corresponds exactly to the ratio expected by the double-backpropagation algorithm. But when the ConvNet has batchnorm layers, this ratio goes up to approximately 1:30.

Note that these ratios are independent of the number of layers.
For instance, the output of the following code with 3 conv layers (n_layers=3) gives:

> No BN and no lambda
> epoch  0 loss 2.30 time 9
> No BN and with lambda
> epoch  0 loss 2.30 time 50
> With BN and no lambda
> epoch  0 loss 2.30 time 12
> With BN and with lambda
> epoch  0 loss 2.30 time 305


```python
import torch                                                                                                                                                                                                                                                                                                                                                                
import torch.nn as nn
import torch.nn.functional as F
import torch.backends.cudnn as cudnn
from torch.autograd import Variable
import time


class Identity(nn.Module):
    def __init__(self, *args):
        super(Identity, self).__init__()

    def forward(self, x): 
        return x

class Layer(nn.Module):
    def __init__(self, in_planes, planes, BN):
        super(Layer, self).__init__()
        self.bn = BN(planes)
        self.conv = nn.Conv2d(in_planes, planes, 3, padding=1, bias=False)
        self.act = nn.ReLU()

    def forward(self, x): 
        return self.act(self.bn(self.conv(x)))

class Net(nn.Module):
    def __init__(self, BN, n_layers=3):
        super(Net, self).__init__()
        self.in_planes = 3 
        self.layers = self._make_layers(Layer,  64, n_layers, BN) 
        self.linear = nn.Linear(64, 10) 

    def _make_layers(self, block, planes, num_blocks, BN):
        strides = [1] + [1] * (num_blocks - 1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_planes, planes, BN))
            self.in_planes = planes
        return nn.Sequential(*layers)

    def forward(self, x): 
        out = self.layers(x)
        out = F.avg_pool2d(out, out.size(2))
        out = out.view(out.size(0), -1) 
        return self.linear(out)

def do_epoch(net, criterion, optimizer, lam):
    net.train()
    aggr_loss = count = 0 
    for _ in range(1000):
        inputs = Variable(torch.cuda.FloatTensor(128,3,32,32).normal_(), requires_grad=True)
        targets = Variable(torch.LongTensor(128).random_(0, 10).cuda())
        outputs = net(inputs)

        optimizer.zero_grad()
        loss = criterion(outputs, targets)
        loss.backward(create_graph=(lam > 0)) 

        # gradient penalty
        if lam > 0:
            gpenalty = inputs.grad.view(inputs.size(0), -1).add(1e-5).norm(1, 1).mean()
            (lam * gpenalty).backward()

        optimizer.step()

        count += 1
        aggr_loss += loss.data[0]

    return aggr_loss / count

def main(net, lam):
    net.cuda()
    cudnn.benchmark = True

    criterion = torch.nn.CrossEntropyLoss()
    criterion.cuda()

    optimizer = torch.optim.SGD(net.parameters(), lr=.001, momentum=0.9)

    for epoch in range(1):
        time_start = time.time()

        loss = do_epoch(net, criterion, optimizer, lam)

        print("epoch %2d loss %.2f time %d" % (epoch, loss, time.time()-time_start))

n_layers = 3 
BN = Identity
print('No BN and no lambda')
main(Net(BN, n_layers), 0.) 

print('No BN and with lambda')
main(Net(BN, n_layers), .001)

BN = nn.BatchNorm2d
print('With BN and no lambda')
main(Net(BN, n_layers), 0.)

print('With BN and with lambda')
main(Net(BN, n_layers), .001)
```

**Comment by fmassa** [comment]
This might probably happen because double backprop in `BatchNorm` is [entirely implemented in python](https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/batchnorm_double_backwards.py) for the moment, and does a lot of autograd operations, while double backprop [in `Conv` is written in C++](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/convolution.cpp#L454-L632) and calls into the (optimized) `Conv` operations (written using cudnn).

There are plans to move some operations to C++, but I'm not sure about the time tables for double backprop for batch norm.

**Comment by gchanan** [comment]
I think @fmassa is correct.  I'll take a look at this when I get a chance (Probably not for a couple of weeks).

**Comment by zou3519** [comment]
I think BatchNorm has a [double backward function in Aten](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/batch_normalization.cpp#L170) that should be called... but I'll check if it is

**Comment by ezyang** [comment]
@zou3519 I already have a port of the batchnorm double backwards. Let me slice it out of my diff and post it.

**Comment by ngimel** [comment]
But unless there are some jit-fused kernels there, it is still an awful lot of kernels, regardless of whether python or aten function is called, right? https://github.com/pytorch/pytorch/blob/925208af72cd956e5cee0a965da632c7985d6c74/torch/nn/_functions/thnn/batchnorm_double_backwards.py#L23-L95

**Comment by ezyang** [comment]
Here it is https://github.com/pytorch/pytorch/pull/4027

And yes, lack of fused kernels is still going to kill you. The JIT fuser should help (although there is some faking up of variables from tensors which may thwart the tracer if applied naively.)

**Comment by lucasb-eyer** [comment]
@cjsg have you tried #4027? What performance do you get with it?

**Comment by cjsg** [comment]
I don't have access to the exact same device than in October 2017 anymore, but, using the same GPU model and pytorch 0.4.0a0+363de58, I now get:
```
No BN and no lambda
epoch  0 loss 2.30 time 11
No BN and with lambda
epoch  0 loss 2.30 time 34
With BN and no lambda
epoch  0 loss 2.30 time 12
With BN and with lambda
epoch  0 loss 2.30 time 64
```

**Comment by fartashf** [comment]
Is this issue fully resolved? I'm doing jacobian vector product on `vgg19` vs `vgg19_bn` and it is very slow on `vgg19_bn`. That is aside from a memory issue with `vgg19_bn` that forces me to use a batch_size of 8.

The output I get from the code below:
```
vgg19 forward: 0.0225s
vgg19 grad 1: 0.0212s
vgg19 grad 2: 0.0081s
vgg19_bn forward: 0.0249s
vgg19_bn grad 1: 0.0259s
vgg19_bn grad 2: 0.7063s
```

Here is the code I use to test:

```python
from torch.autograd import Variable
import torch
import torch.nn.functional as F
import torchvision.models as models
import time


def dbp(model):
    retain_graph = False
    W = model.parameters()
    batch_size = 8
    x = Variable(torch.randn(batch_size, 3, 224, 224)).cuda()
    t = Variable(torch.ones(batch_size).long()).cuda()

    y = model(x)
    tic = time.time()
    loss_ex = F.nll_loss(y, t, reduce=False)/batch_size
    loss = loss_ex.sum()
    t0 = time.time()-tic

    tic = time.time()
    v = Variable(torch.ones_like(loss_ex.data), requires_grad=True).cuda()
    grad_params = torch.autograd.grad(loss_ex, W, v,
                                      retain_graph=retain_graph)
    t1 = time.time()-tic
    m = [torch.ones_like(g) for g in grad_params]
    tic = time.time()
    jvp = torch.autograd.grad(grad_params, v, m,
                              retain_graph=retain_graph)[0]
    t2 = time.time()-tic
    del loss, jvp
    return t0, t1, t2


def test_dbp(arch):
    t0 = 0
    t1 = 0
    t2 = 0
    num = 10
    model = models.__dict__[arch]()
    model.cuda()
    for i in range(num):
        ttt = dbp(model)
        t0 += ttt[0]
        t1 += ttt[1]
        t2 += ttt[2]

    print('%s forward: %.4fs' % (arch, t0/num))
    print('%s grad 1: %.4fs' % (arch, t1/num))
    print('%s grad 2: %.4fs' % (arch, t2/num))

test_dbp('vgg19')
test_dbp('vgg19_bn')
```

**Comment by ngimel** [comment]
No, it's not resolved, it requires custom double backward kernels for batch norm.

**Comment by fartashf** [comment]
Any specific plans for custom kernels?

**Comment by fmassa** [comment]
I don't think there are specific plans to optimize it yet. But from a quick glance at the code, it looks like this is the perfect candidate for applying jit compilation

**Comment by ngimel** [comment]
There are a lot of reductions there (sum_exclude_dim1) that are not supported by jit?
